src/storage/src/source/reclock/compat.rs

```diff
@@ -0,0 +1,436 @@
+                    // happened during compare_and_append.
+                    // Vector holding the result of the translation
+                    let mut native_frontier_updates = vec![];
+
+                    // First, we will sort the updates by time to be able to iterate over the groups
+                    updates.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+
+                    // This is very subtle. An empty collection of native Antichain elements
+                    // represents something different than an empty collection of compat
+                    // OffsetAntichain elements. The former represents the empty antichain and the
+                    // latter the Antichain containing the minimum element. Therefore we need to
+                    // always synthesize a minimum timestamp element that happens once at the as_of
+                    // before processing any updates from the shard.
+                    if !std::mem::replace(&mut self.minimum_produced, true) {
+                        let mut first_ts = mz_repr::Timestamp::minimum();
+                        first_ts.advance_by(self.as_of.borrow());
+                        native_frontier_updates.push((Partitioned::minimum(), first_ts, 1));
+                    }
+
+                    // Then, we will iterate over the group of updates in time order and produce
+                    // the OffsetAntichain at each point, convert it to a normal Antichain, and
+                    // diff it with current Antichain representation.
+                    for (ts, updates) in &updates.into_iter().group_by(|update| update.1) {
+                        let prev_native_frontier =
+                            Antichain::from(self.compat_source_upper.frontier());
+                        native_frontier_updates.extend(
+                            prev_native_frontier
+                                .into_iter()
+                                .map(|src_ts| (src_ts, ts, -1)),
+                        );
+
+                        self.compat_source_upper.update_iter(
+                            updates
+                                .into_iter()
+                                .map(|(binding, _ts, diff)| (binding, diff)),
+                        );
+                        let new_native_frontier =
+                            Antichain::from(self.compat_source_upper.frontier());
+
+                        native_frontier_updates.extend(
+                            new_native_frontier
+                                .into_iter()
+                                .map(|src_ts| (src_ts, ts, 1)),
+                        );
+                    }
+                    // Then, consolidate the native updates and we're done
+                    consolidation::consolidate_updates(&mut native_frontier_updates);
+
+                    // Finally, apply the updates to our local view of the native frontier
+                    self.native_source_upper.update_iter(
```

Comment:
i think it may also be worth documented (if we have time before we delete this) why we update these in `next` but not `compare_and_append`; that adds an implicit require on how this api is sued